### DISCUSSION TITLE
401 when connecting data importer

### DISCUSSION THREAD

**Discussion by nimjor** [firefly-iii/firefly-iii#5902] [Firefly III]
### Support guidelines

- [X] I've read the [support guidelines](https://github.com/firefly-iii/firefly-iii/blob/main/.github/support.md)

### I've found a bug and checked that ...

- [X] ... [the documentation](https://docs.firefly-iii.org/) does not mention anything about my problem
- [X] ... there are no open or closed issues that are related to my problem
- [X] ... it's [definitely me, not you](https://github.com/firefly-iii/firefly-iii/blob/main/.github/its_you_not_me.md)

### Description

Seems like a great app, but I haven't been able to get going with it much yet because I can't get the data importer authorized.  I've read the documentation on how to do it (i.e. _un-checking_ the "confidential" option when creating a new OAuth client), verified the callback URL, etc.  No matter what I've tried, after I click "Authorize" (this step _does_ show the correct client name as configured in Firefly III), I receive a 401 when it redirects me back to the FIDI callback URL.  I've made sure my URLs are set correctly in .fidi.env.  I've also tried ditching the client ID and using a personal access token instead, but that doesn't work either.

### Expected behaviour

When following the steps as documented, authentication should succeed.

### Debug information

Debug info: 
```
Debug information generated at 2022-03-15 22:08:11 US/Central for Firefly III Data Importer version **0.9.0**.

| Scope | Version |
| --- | --- |
| FIDI | 0.9.0 |
| Build | 97, 21-02-2022 20:13:33 CET |
| PHP | 8.0.13 |
| Host | Linux |

| System info | Value |
| --- | --- |
| Using docker? | true |
| System TZ | US/Central |
| Browser TZ | America/Chicago |
| App environment | local |
| App debug mode | false |
| App cache driver | file |
| App logging | warning, stack |
| Display errors | Off |
| Error reporting | ALL errors |
| Interface | apache2handler |
| BCscale | 12 |
| Trusted proxies (.env) | ** |

| User info | Value |
| --- | --- |
| User agent | Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.3 Safari/605.1.15 |
```

Additional info:
```
Truncated from this point <----|
#6 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Middleware/SubstituteBindings.php(50): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#7 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Routing\\Middleware\\SubstituteBindings->handle()
#8 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/VerifyCsrfToken.php(78): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#9 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\VerifyCsrfToken->handle()
#10 /var/www/html/vendor/laravel/framework/src/Illuminate/View/Middleware/ShareErrorsFromSession.php(49): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#11 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\View\\Middleware\\ShareErrorsFromSession->handle()
#12 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(121): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#13 /var/www/html/vendor/laravel/framework/src/Illuminate/Session/Middleware/StartSession.php(64): Illuminate\\Session\\Middleware\\StartSession->handleStatefulRequest()
#14 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Session\\Middleware\\StartSession->handle()
#15 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/AddQueuedCookiesToResponse.php(37): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#16 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\AddQueuedCookiesToResponse->handle()
#17 /var/www/html/vendor/laravel/framework/src/Illuminate/Cookie/Middleware/EncryptCookies.php(67): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#18 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Cookie\\Middleware\\EncryptCookies->handle()
#19 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#20 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(723): Illuminate\\Pipeline\\Pipeline->then()
#21 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(698): Illuminate\\Routing\\Router->runRouteWithinStack()
#22 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(662): Illuminate\\Routing\\Router->runRoute()
#23 /var/www/html/vendor/laravel/framework/src/Illuminate/Routing/Router.php(651): Illuminate\\Routing\\Router->dispatchToRoute()
#24 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(167): Illuminate\\Routing\\Router->dispatch()
#25 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(128): Illuminate\\Foundation\\Http\\Kernel->Illuminate\\Foundation\\Http\\{closure}()
#26 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#27 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ConvertEmptyStringsToNull.php(31): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#28 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ConvertEmptyStringsToNull->handle()
#29 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TransformsRequest.php(21): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#30 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/TrimStrings.php(40): Illuminate\\Foundation\\Http\\Middleware\\TransformsRequest->handle()
#31 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\TrimStrings->handle()
#32 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/ValidatePostSize.php(27): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#33 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\ValidatePostSize->handle()
#34 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Middleware/PreventRequestsDuringMaintenance.php(86): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#35 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Foundation\\Http\\Middleware\\PreventRequestsDuringMaintenance->handle()
#36 /var/www/html/vendor/fruitcake/laravel-cors/src/HandleCors.php(38): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#37 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Fruitcake\\Cors\\HandleCors->handle()
#38 /var/www/html/vendor/laravel/framework/src/Illuminate/Http/Middleware/TrustProxies.php(39): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#39 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(167): Illuminate\\Http\\Middleware\\TrustProxies->handle()
#40 /var/www/html/vendor/laravel/framework/src/Illuminate/Pipeline/Pipeline.php(103): Illuminate\\Pipeline\\Pipeline->Illuminate\\Pipeline\\{closure}()
#41 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(142): Illuminate\\Pipeline\\Pipeline->then()
#42 /var/www/html/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(111): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter()
#43 /var/www/html/public/index.php(77): Illuminate\\Foundation\\Http\\Kernel->handle()
#44 {main}
"} 
[2022-03-15 21:57:17] local.ERROR: Client exception when decoding response: Client error: `POST https://firefly.mydomain.com/oauth/token` resulted in a `401 Unauthorized` response:
<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></cen (truncated...)
  
[2022-03-15 21:57:17] local.ERROR: Response from server: "<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></center>
<hr><center>openresty</center>
</body>
</html>
"  
[2022-03-15 22:05:07] local.ERROR: Client exception when decoding response: Client error: `POST https://firefly.mydomain.com/oauth/token` resulted in a `401 Unauthorized` response:
<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></cen (truncated...)
  
[2022-03-15 22:05:07] local.ERROR: Response from server: "<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></center>
<hr><center>openresty</center>
</body>
</html>
"  
[2022-03-15 22:06:44] local.EMERGENCY: I am a EMERGENCY message.  
[2022-03-15 22:06:44] local.ALERT: I am a ALERT message.  
[2022-03-15 22:06:44] local.CRITICAL: I am a CRITICAL message.  
[2022-03-15 22:06:44] local.ERROR: I am a ERROR message.  
[2022-03-15 22:06:44] local.WARNING: I am a WARNING message.  
[2022-03-15 22:08:05] local.ERROR: Client exception when decoding response: Client error: `POST https://firefly.mydomain.com/oauth/token` resulted in a `401 Unauthorized` response:
<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></cen (truncated...)
  
[2022-03-15 22:08:05] local.ERROR: Response from server: "<html>
<head><title>401 Authorization Required</title></head>
<body>
<center><h1>401 Authorization Required</h1></center>
<hr><center>openresty</center>
</body>
</html>
"  
[2022-03-15 22:08:11] local.EMERGENCY: I am a EMERGENCY message.  
[2022-03-15 22:08:11] local.ALERT: I am a ALERT message.  
[2022-03-15 22:08:11] local.CRITICAL: I am a CRITICAL message.  
[2022-03-15 22:08:11] local.ERROR: I am a ERROR message.  
[2022-03-15 22:08:11] local.WARNING: I am a WARNING message.  
```

### Steps to reproduce

1. Configure docker-compose.yml, .env, and .fidi.env according to documentation, accounting for necessary settings for reverse proxies.
2. Start the services (docker-compose up -d)
3. Create Firefly III account and log in
4. Navigate in separate browser tab to data importer to get callback URL.
5. Use data importer callback URL to set up new OAuth client in Firefly III
6. Back over at data importer, enter the newly generated client ID (number, like 43) and the Firefly III URL, and submit the form.
7. Step 6 sends us to https://[firefly-iii-url]/oauth/authorize?client_id=43&redirect_uri=[url-encoded fidi url]/callback&response_type=code&scope=&state=[mystate]&code_challenge=[mycodechallenge]&code_challenge_method=S256
8. Clicking the Authorize button on the Firefly III page from step 7 redirects back to the FIDI callback URL with the OAuth code and state passed as query params, but showing that a 401 was received:
```
Client error: `POST https://firefly.mydomain.com/oauth/token` resulted in a `401 Unauthorized` response: <html> <head><title>401 Authorization Required</title></head> <body> <center><h1>401 Authorization Required</h1></cen (truncated...)
```

### Additional info

I'm running the latest of both images (firefly-iii and fidi) in docker, based on the docker-compose example provided.  I've modified the relevant variables for my environment (nginx reverse proxy with SSL, both apps on their own subdomains).

I'm happy to provide more information as needed.

**Comment by nimjor** [comment]
I forgot to mention this--it seems strange that FIDI would prompt for client ID and Firefly III url when these are passed in as part of the .fidi.env file (yes my docker-compose.yml points my **fidi** service **env_file** to `.fidi.env`).  I've tried every combination I can think of with the environment variables, followed by `docker-compose restart` each time, to no avail.

**Comment by JC5** [comment]
Clear your cookies or press reauthenticate.

**Reply by nimjor** [reply]
Sorry, I forgot that piece too--I tried many things.  Different browsers (had never used before to access firefly/fidi), private/incognito sessions, the "reauthenticate" and "start over" buttons, etc.  I've tried once again, clearing all cookies, pressing reauthenticate--still nothing.  Are there logs that would shed some more light on this specific issue, something I can look for in configurations?  Would you like a short screen recording or screenshots?

**Comment by JC5** [comment]
Are you using Portainer, by any chance?

**Reply by nimjor** [reply]
Yes! I didn’t use it to configure the services, only to pull the images initially. The rest of the setup I did manually via CLI. If Portainer causes problems with Firefly, I’m don’t really use it at all most of the time and could probably part with it if there isn’t a workaround.

**Reply by JC5** [reply]
I know some people have trouble with Portainer starting and running the containers, as it doesn't seem to let them set environment variables through config files like you are doing right now. Since you're also complaining about not being able to set the URL, I thought this could be the cause.

I doubt Portainer is really buggy but I do see everybody running into the same issue.

**Reply by nimjor** [reply]
It's strange, looking at the config via Portainer I see that FIREFLY_III_CLIENT_ID is null, but in my .fidi.env it is set (and has been this whole time, various values as I've created and deleted OAuth clients in Firefly III).  All the other values seem to have been picked up correctly from .fidi.env.  Do you see anything wrong with this part of the file?  I'm just using the example compose file.

Portainer:
<img width="856" alt="Screen Shot 2022-03-17 at 11 34 39 AM" src="https://user-images.githubusercontent.com/76883392/158850521-9eda05f3-1feb-4e8d-85e9-07381d814542.png">

docker-compose.yml:
<img width="435" alt="Screen Shot 2022-03-17 at 11 36 18 AM" src="https://user-images.githubusercontent.com/76883392/158850646-a0941bb6-34d0-47c0-8884-0a76e2fed333.png">

**Reply by JC5** [reply]
They seem to be correct. I'm not sure why Portainer would not display them. Do you use `docker-compose up` to run?

**Reply by nimjor** [reply]
Correct, except running detached.  It appears it's not so much that Portainer isn't displaying the value, rather it is indeed null:
<img width="490" alt="Screen Shot 2022-03-17 at 11 52 31 AM" src="https://user-images.githubusercontent.com/76883392/158852898-01165429-7615-4a9a-9cb3-9d51e9f72adb.png">

**Reply by nimjor** [reply]
Is there any way something in the FIDI app could be _unsetting_ that particular value upon startup?  I've tried wrapping the client ID value in quotes, also, to no avail.

**Reply by JC5** [reply]
No, especially not the environment vars. Is there a `.env` file in the same directory, on the container itself? That shouldn't overrule anything but still.

**Reply by nimjor** [reply]
I've solved the problem!  After deleting the existing containers and volumes, rebuilding, setting up my account in Firefly and adding an OAuth client (all before navigating to or logging in to FIDI), then stopping everything, setting that client ID in .fidi.env, starting everything and navigating to my FIDI site, it seemed to pick up the env variable.  I wasn't getting the page prompting me to enter the Firefly URL or client ID anymore, and instead it immediately took me to the Firefly auth page to authorize the FIDI client.  However, upon submitting, I got a 502, which I typically have only seen from my nginx proxy when the host is down.  After looking in the nginx error log for that host, I found this error "upstream sent too big header while reading response header from upstream", for which I found a solution on [this issue](https://github.com/firefly-iii/firefly-iii/issues/3892).  I've just successfully completed my first CSV import.  Thank you for your help, I'm excited to explore the app!

**Reply by JC5** [reply]
Glad to hear it works!
